### ISSUE TITLE
[audio] Anomaly in Streets of Rage Boss music

### ISSUE THREAD

**Issue by zenzero-2001** [MiSTer-devel/Genesis_MiSTer#130]
Hi. Sorry if this issue is a duplicate. Part of the music for the Streets of Rage Boss track sounds wrong / different. I have compared to Nuked via Genesis Plus GX and to the version from the 16-bit audiophile project. The problem is more obvious with ym3438 and no filter but also exists with ym2612 and model 1 filter. Anomaly occurs between 79 and 84 seconds from the beginning of the track. Using genesis core 200502.

Attached zip of relevant excerpts:

[mister_sor_bug.zip](https://github.com/MiSTer-devel/Genesis_MiSTer/files/4672028/mister_sor_bug.zip)

**Comment by SegaSnatcher** [comment]
I have a pretty good ear for this kind of stuff and I literally do not hear what you are talking about.  In fact the Audiophile Project track and MiSTer Model 1 filter sounds near identical.  Can you please elaborate what differences you are hearing?  

FWIW,  Streets of Rage relied on YM2612's Ladder effect.  Using a YM3438 variant will make it sound somewhat different.

**Comment by Kitrinx** [comment]
Please provide real recordings from original hardware to demonstrate the difference. Emulators, while sometimes helpful, aren't great for reference examples.

**Comment by SegaSnatcher** [comment]
> Please provide real recordings from original hardware to demonstrate the difference. Emulators, while sometimes helpful, aren't great for reference examples.

The Audiophile Project clip provided is real hardware.

**Comment by zenzero-2001** [comment]
Hi. Thanks for the comments. I am attaching a screenshot from Audacity highlighting the problem area.

![sor_boss_mister_2612_m1filter](https://user-images.githubusercontent.com/24979714/83350955-6d412e00-a32f-11ea-968b-23b3d38af584.jpg)

**Comment by SegaSnatcher** [comment]
That doesn't tell us anything, again those look pretty much identical.  What is it that you are actually hearing?  You are not being specific enough.  You keep mentioning "Problem" but you won't specify what it is.

**Comment by SegaSnatcher** [comment]
I think I finally hear what you might be talking about.  Is it the white noise volume swell at the end?  On MiSTer it seems to start somewhat earlier than real hardware?

**Comment by zenzero-2001** [comment]
Hi. Sorry for not being specific. I did not think it would be difficult for other people to hear the issue. :-) I am listening with an external DAC and amplifier into headphones which may make it easier for me to hear.

It is not the white noise volume swell, nor is it the timing that is the issue.

The problem is to do with the beat, I'm not sure how to describe it. On the MiSTer version there is an after-tone accompanying each bass hit.

The Nuked and audiophile versions sound how I would expect. But, maybe MiSTer is correct and the others are not? Unfortunately, I do not have a real Genesis / Megadrive to compare to.

**Comment by James-F2** [comment]
The tracks are so similar they cancel out when flipping the phase of one of them.
Are you sure it's not a problem with the DAC?

![sega](https://user-images.githubusercontent.com/31185814/83354948-cbd4cf00-a364-11ea-89da-ce60ad39031a.png)

**Comment by zenzero-2001** [comment]
Hi. I tried the DAC inside my laptop and the issue is still audible.

I did not describe what I hear very well. The anomaly overlays the beat hits, definitely the first two, not sure about the third. It sounds a bit like a looseness to the beat or a vibration / resonance.

**Comment by Ace9921** [comment]
I've taken a sample from my Japanese MegaDrive model 1 (VA5 revision) and can hear what seems like an improperly-generated metallic sound in the bass hits at the highlighted section on MiSTer.  I've attached the sample in question here:
[Streets of Rage VA5 MegaDrive model 1.zip](https://github.com/MiSTer-devel/Genesis_MiSTer/files/4708423/Streets.of.Rage.VA5.MegaDrive.model.1.zip)
Compare this to the MiSTer YM2612 sample.  At around the 8.2 second mark is when the difference is audible.

**Comment by zenzero-2001** [comment]
Thanks Ace9921, I was beginning think I was imagining it. :-)

"Metallic" sound is definitely a better description.

**Comment by bfbiii** [comment]
In case anyone is having trouble hearing this, here are 8 samples of OEM in that section followed by 8 samples of MiSTer. The squeak or metallic sound should stick out this way.

[https://www.dropbox.com/s/3o359dvbbqfzrz4/8%20Samples%20OEM--8%20Samples%20MiSTer.wav?dl=0](url)

**Comment by SegaSnatcher** [comment]
> In case anyone is having trouble hearing this, here are 8 samples of OEM in that section followed by 8 samples of MiSTer. The squeak or metallic sound should stick out this way.
> 
> [https://www.dropbox.com/s/3o359dvbbqfzrz4/8%20Samples%20OEM--8%20Samples%20MiSTer.wav?dl=0](url)

Your link doesn't work for me.  Takes me to the front page of Genesis MiSTer dev.

**Comment by bfbiii** [comment]
I edited it, does it work now?

**Comment by SegaSnatcher** [comment]
> I edited it, does it work now?

Yes it works now.

**Comment by bfbiii** [comment]
That is zenzero-2001's recording and Ace's recording. I just cut it to create a repeated pattern for each--so folks could hear. ABXing with analogous patterns is far easier.

If you hear the squeak or metallic sound kinda between the beats, then you hear it.

**Comment by jotego** [comment]
Hi guys, I am the author of JT12, which is forked inside MiSTer. If you provide me with a VGM file of the track that has the problem I can work on it. There is a large repository of VGM files [here](https://project2612.org/list.php). Just point me to the one I need to check.
I can then verify the problem in the original JT12 repository, correct it if needed and alert the curators of this fork about the update.

**Comment by bfbiii** [comment]
The track in question is "12 - Attack the Barbarian"

https://www.dropbox.com/sh/e49ilj0a4oip194/AADAZrbQQUZDbe6UD4HW83ona?dl=0

At VGM 60Hz, the bass hits are 2m43.247-2m74.791

The moment before the second bass hit is the sound.

**Comment by donluca** [comment]
>
> EDIT: I just learned the 16-Bit Audiophile Project uses VGM for playback to record audio from a real Genesis, so might not be a good reference to use after all since VGM playback timings might not be totally accurate.
> 
> ![unknown (4)](https://user-images.githubusercontent.com/36483667/83352551-d3b75380-a311-11ea-9d6b-7e8be0f3654d.png)

Our Hardware VGM Player is 100% accurate, down to each single cycle and thanks to MDFourier we've been able to confirm this.

**Comment by jotego** [comment]
Sadly, most issues are more blunt than a cycle problem. So the VGM file should suffice. Give me a few days to look at it.

**Comment by donluca** [comment]
I've inspected our release and it turned out to be one of the older ones which need to be remastered. Back then we were using a PAL Mega Drive with a 60hz switch and swapped crystal but discovered later that it wasn't enough.
That's the reason why our file has an altered pitch and slightly slower speed.
Remastered releases don't have this issue, nor releases after November 2014,

**Comment by bootsector** [comment]
Signing up :tophat:

**Comment by jotego** [comment]
Ok, I finally got down to looking at this today.

I checked out the samples that @bfbiii had kindly prepared. I do hear a slight difference. Upon inspection, there are some slight differences in the waveform:

![comparison](https://user-images.githubusercontent.com/1863036/84791230-9e3d8600-aff2-11ea-849c-2b15813fef07.png)

The top one appears in the first half of @bfbiii samples, the bottom corresponds to the second half. I believe that the the top is the one recorded from one Megadrive and the bottom from a MiSTer.

Here are the audacity files so you can compare by yourself. The two signals are virtually  identical except for that filtering. 
[8 sample comparison - audacity.zip](https://github.com/MiSTer-devel/Genesis_MiSTer/files/4787357/8.sample.comparison.-.audacity.zip)

I know MiSTer has some filtering options and also a PCM enhancer option. I think these options may have something to do. Also MiSTer audio backend will add some colour to the sound too. The analogue output has different filters from those of any specific hardware (Megadrive, SNES, etc) and the digital section didn't properly deal with upscaling issues the last time I saw it. Having said that, the rough interpolator used in the digital section apparently was enough to stop complains from users.

I have reviewed the signal path for PCM samples inside the JT12 sound module and it does match the original one (we do know how the original one was). So if you disable the PCM enhancer the output from JT12 will match the original YM2612 output except for the difference in the DAC. You can try enabling the ladder effect to see if that matches your expectation.

PCM data actually goes across YM2612 without processing so as long as the Z80 gets it in at the right time, there is not much else involved. I don't think that the Z80 in MiSTer will fail to do things at the right time.

In summary, JT12 output should be identical. There are a number of filtering options that can be used to change the sound colour a bit. If someone is interested in matching FPGA and Genesis outputs, then a spectrum sweeper could be built in order to make a spectrum analysis of MiST(er) vs Genesis response and add an equalizer for them. That is independent of JT12 so I don't think you need me for that.

**Comment by donluca** [comment]
Thank you Jose, your work and dedication is really an example and inspiration to all of us.

Slightly unrelated, but I've confirmed that VGM has some tiny timing issues due to rounding errors in the conversion to 44.1Khz which is the frequency used by the format itself.
I don't want to pollute this thread, but I hope to have time soon to make a proper post and explanation on my own project's website.

**Comment by bfbiii** [comment]
@jotego Thanks! As for FPGA and Genesis outputs with sweeps, Artemio has made MDFourier to do that
http://junkerhq.net/MDFourier/
and generally the MiSTer Genesis core with Model 1 filtering is pretty accurate to OEM. MDFourier does not test the 6th channel DAC feature of the YM2612 with a sweep, so if that feature is in play here, that would be interesting. Related: the MiSTer framework is getting a whole new set of filter options with special attention towards anti-aliasing and DC off-set, and that may or may not knock out this anomaly.
